metabase/metabase
# support timeline events on the new combo chart

Related to https://github.com/metabase/metabase/issues/35781

### Description

This PR adds the support of timeline events to the dynamic combo chart. It includes:
- Showing timeline events within the visible x-axis range
- Highlight selected timeline events
- Merge timeline events that are too close to each other

We intentionally group timeline events by the beginning of a bucket based on the x-axis unit. For example, if a chart has monthly bucketing, December 1 and December 31 will be rendered as one line with the count of events in the group on the right.

This is the first piece of functionality so far that required knowing chart width for the right computation of the option object which led to a slight refactoring of the cartesian chart model to include there everything that is needed for measuring chart elements such as ticks width. This is needed to get the actual boundary of the chart to calculate distance between timeline events in pixels so we could merge overlapping ones. It whould also be more convenient now with axis models of `YAxisModel` type instead of having multiple separate optional properties for axis split, axis columns.

The following items will be added further:
- Timeline events tooltip
- Compute timeline events binning for native queries by inspecting data

### How to verify

- Ensure timeline events are visible in the query builder
- Ensure timeline events selected in the sidebar are highlighted on the chart
- Ensure timeline events are grouped by the dimension bucket
- Ensure timeline events are not overlapping and become grouped in this case

### Demo

https://github.com/metabase/metabase/assets/14301985/5a9bea8e-a53a-407f-86e9-a5898b7aaaf4

### Checklist
